File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/ApprovalGuard.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Text.RegularExpressions

module ApprovalGuard =
    type ApprovalDecision =
        | Approved
        | Denied of string

    type ApprovalRequest = {
        Action: string
        Reason: string
        Payload: string
    }

    type ApprovalPrompt = ApprovalRequest -> Async<ApprovalDecision>

    let private riskyCommandPattern =
        Regex(@"\b(rm\s+-rf|del\s+/f|format\b|shutdown\b|reboot\b|mkfs\b|diskpart\b|sudo\b)\b", RegexOptions.IgnoreCase ||| RegexOptions.Compiled)

    let private riskyEditPattern =
        Regex(@"(?i)(password|secret|token|private[_-]?key|connectionstring)", RegexOptions.Compiled)

    let isHighRiskCommand (command: string) =
        not (String.IsNullOrWhiteSpace(command)) && riskyCommandPattern.IsMatch(command)

    let isHighRiskFileEdit (path: string) (contentPreview: string) =
        let sensitiveFile =
            not (String.IsNullOrWhiteSpace(path)) &&
            (path.EndsWith(".env", StringComparison.OrdinalIgnoreCase)
             || path.EndsWith("secrets.json", StringComparison.OrdinalIgnoreCase)
             || path.Contains("credential", StringComparison.OrdinalIgnoreCase))

        sensitiveFile || (not (String.IsNullOrWhiteSpace(contentPreview)) && riskyEditPattern.IsMatch(contentPreview))

    let createConsolePrompt () : ApprovalPrompt =
        fun request ->
            async {
                printfn "\nApproval required for action: %s" request.Action
                printfn "Reason: %s" request.Reason
                printfn "Payload: %s" request.Payload
                printf "Allow? (y/N): "
                let input = Console.ReadLine()
                if not (isNull input) && input.Trim().Equals("y", StringComparison.OrdinalIgnoreCase) then
                    return Approved
                else
                    return Denied "User rejected action."
            }

    let enforceApproval (prompt: ApprovalPrompt) (request: ApprovalRequest) : Async<Result<unit, string>> =
        async {
            let! decision = prompt request
            match decision with
            | Approved -> return Ok ()
            | Denied reason -> return Error reason
        }

    let requireCommandApproval (prompt: ApprovalPrompt) (command: string) : Async<Result<unit, string>> =
        async {
            if isHighRiskCommand command then
                let request = {
                    Action = "execute_command"
                    Reason = "Command classified as high-risk."
                    Payload = command
                }
                return! enforceApproval prompt request
            else
                return Ok ()
        }

    let requireFileEditApproval (prompt: ApprovalPrompt) (path: string) (contentPreview: string) : Async<Result<unit, string>> =
        async {
            if isHighRiskFileEdit path contentPreview then
                let request = {
                    Action = "edit_file"
                    Reason = "File edit classified as sensitive/high-risk."
                    Payload = sprintf "%s | %s" path contentPreview
                }
                return! enforceApproval prompt request
            else
                return Ok ()
        }

File: implementations/csharp/src/Skight.AgentPlatform/ApprovalGuard.cs
using System;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Skight.AgentPlatform
{
    public class ApprovalRequest
    {
        public string Action { get; set; } = string.Empty;
        public string Reason { get; set; } = string.Empty;
        public string Payload { get; set; } = string.Empty;
    }

    public static class ApprovalGuard
    {
        private static readonly Regex RiskyCommandPattern = new(@"\b(rm\s+-rf|del\s+/f|format\b|shutdown\b|reboot\b|mkfs\b|diskpart\b|sudo\b)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RiskyEditPattern = new(@"(?i)(password|secret|token|private[_-]?key|connectionstring)", RegexOptions.Compiled);

        public static bool IsHighRiskCommand(string command)
        {
            return !string.IsNullOrWhiteSpace(command) && RiskyCommandPattern.IsMatch(command);
        }

        public static bool IsHighRiskFileEdit(string path, string contentPreview)
        {
            var sensitiveFile = !string.IsNullOrWhiteSpace(path) && 
                (path.EndsWith(".env", StringComparison.OrdinalIgnoreCase) || 
                 path.EndsWith("secrets.json", StringComparison.OrdinalIgnoreCase) || 
                 path.Contains("credential", StringComparison.OrdinalIgnoreCase));

            return sensitiveFile || (!string.IsNullOrWhiteSpace(contentPreview) && RiskyEditPattern.IsMatch(contentPreview));
        }

        public static async Task<bool> RequestConsoleApprovalAsync(ApprovalRequest request)
        {
            Console.WriteLine($"\nApproval required for action: {request.Action}");
            Console.WriteLine($"Reason: {request.Reason}");
            Console.WriteLine($"Payload: {request.Payload}");
            Console.Write("Allow? (y/N): ");
            
            // In a real app this would read asynchronously, but for console this is synchronous mostly
            var input = await Task.Run(() => Console.ReadLine());
            return !string.IsNullOrEmpty(input) && input.Trim().Equals("y", StringComparison.OrdinalIgnoreCase);
        }

        public static async Task EnforceCommandApprovalAsync(string command, Func<ApprovalRequest, Task<bool>> prompter)
        {
            if (IsHighRiskCommand(command))
            {
                var request = new ApprovalRequest
                {
                    Action = "execute_command",
                    Reason = "Command classified as high-risk.",
                    Payload = command
                };
                if (!await prompter(request)) throw new UnauthorizedAccessException("User rejected action.");
            }
        }

        public static async Task EnforceFileEditApprovalAsync(string path, string contentPreview, Func<ApprovalRequest, Task<bool>> prompter)
        {
            if (IsHighRiskFileEdit(path, contentPreview))
            {
                var request = new ApprovalRequest
                {
                    Action = "edit_file",
                    Reason = "File edit classified as sensitive/high-risk.",
                    Payload = $"{path} | {contentPreview}"
                };
                if (!await prompter(request)) throw new UnauthorizedAccessException("User rejected action.");
            }
        }
    }
}
